### ISSUE TITLE
Consider using the new officially sanctioned clj-msi when installing clojure on Windows

### ISSUE THREAD

**Issue by lread** [DeLaGuardo/setup-clojure#65]
Hiya! 

I am excited to see support for `deps.exe` (as `clojure.exe`) on Windows.
Thanks so much for adding that as a feature!

The current README suggests selecting via:
```
  cmd-exe-workaround: 'latest' # Replaces `clojure` with `deps.clj` on Windows
```

But the source code sez `cmd-exe-workaround` is deprecated https://github.com/DeLaGuardo/setup-clojure/blob/3199aaaa9cf140bcb7a5f36a72d1aa42658ee991/action.yml#L17-L24

But I don't know what a "unified cli input" is.

Let's say I have a matrix build for macOS, Windows, and Linux, how should I select `deps.exe` (as `clojure.exe`) for windows?

**Comment by DeLaGuardo** [comment]
The latest version got a fix to run clojure tools deps on windows the same way as on other platforms. https://github.com/DeLaGuardo/setup-clojure/blob/main/.github/workflows/smoke-tests.yml#L31-L52

**Comment by lread** [comment]
Thanks for the reply!

By cli, you mean the [Clojure official Windows PowerShell module](https://github.com/clojure/tools.deps.alpha/wiki/clj-on-Windows), right?

Can I still use [borkdude's deps.exe (renamed to clojure) on Windows](https://github.com/borkdude/deps.clj)?

Or are you saying that `cli` automatically uses deps.exe on Windows?

(Sorry if the answer is obvious, I'm feeling a little not-so-bright here! 🙂)

**Comment by DeLaGuardo** [comment]
Right, getLatestDepsClj is not used anymore. Instead, when you use input `cli: %version%` for windows platform, it will install official Windows PowerShell Module into the directory visible for powershell and pwsh (default shell for github action runner on windows). That means previously required option `shell: powershell` is not needed anymore. Run step using `clojure` command line will recognise correctly where binary is installed without additional options.

```yml
jobs:
  run-clojure-cli:
    strategy:
      matrix:
        # note all three available platforms
        os: [ubuntu-latest, macOS-latest, windows-latest]

    runs-on: ${{ matrix.os }}

    steps:
      - name: Prepare java
        uses: actions/setup-java@v3
        with:
          distribution: 'zulu'
          java-version: '8'

      - name: Install Clojure CLI
        uses: DeLaGuardo/setup-clojure@9.3
        with:
          cli: 1.10.1.693

      # step uses clojure command is the same across all platforms
      - name: Execute clojure code
        run: clojure -e "(+ 1 1)"
```

**Comment by lread** [comment]
Ah, thanks for the clarification!

I use @borkdude's `deps.exe` as `clojure.exe` to avoid many headaches on Windows.
I think the setup-clojure action was briefly offering this as an option?

Anyway, these days I can always use `bb clojure` on Windows instead.

**Comment by DeLaGuardo** [comment]
`deps.exe` was there as a workaround, yes. But the code to install it is here, so it would be easy to reintroduce it. Do you know some place where I can read about the headaches you mention? I would like to evaluate them and bring back an option to install `deps.exe` or find a way to avoid most problems.

**Comment by borkdude** [comment]
Another reason you might want to use `deps.exe` as `clojure.exe` is shelling out: when shelling out with `ProcessBuilder` or `babashka.process` for that matter, it just works whereas with the Powershell installation you have to write `pwsh.exe -c 'clojure ...'` or equivalent (not exactly sure what the command is, but you get the idea).

**Comment by borkdude** [comment]
Yes, I'm talking about shelling out from e.g. a clojure program, not in the Github Actions "code".

**Comment by DeLaGuardo** [comment]
got it. That sounds like a good reason to add `deps.exe` to the set of tools available to install. thanks!

**Comment by borkdude** [comment]
This environment variable might also be interesting:

https://github.com/borkdude/deps.clj#deps_clj_tools_version

**Comment by lread** [comment]
Yes, what @borkdude said 🙂.
Figuring out argument escaping on Windows is difficult and made more painful when shelling out with Powershell. Capturing the exit code in these cases is also non-obvious for those who aren't Powershell experts.

If/when you do re-add deps.exe as clojure.exe on Windows, it would be extra convenient if I could express I'd like to use this on Windows without repeating a `setup-clojure` with a GitHub Actions `if` for Windows.  I think your `cmd-exe-workaround:` would allow this if it were understood that it overrides `cli:`?

**Comment by DeLaGuardo** [comment]
yes, it will install `clojure` command overtaking the `cli:` input.

Btw. could you share an example of the process that is hard to grok on windows with "official powershell cli"? Then I can add it to the test suite to capture regressions, if any

**Comment by lread** [comment]
The exit code for PowerShell launches directly from GitHub Actions is handled properly, but if you call clojure yourself from a script you have to be aware of [this](https://clojurians.slack.com/archives/CFN4QDHPS/p1653063279747199):

> TIL, (I am powershell naive, so maybe obvious to most of you):
If you run `powershell -command my-command` the exit code from `my-command` is not returned to the caller.
For that you need tag on an `exit $LASTEXITCODE`, like:
`powershell -command "my-command;exit $LASTEXITCODE"`
Kind of important if you are wanting to know about little things like test failures.

I don't have specific examples of escaping woes, but I've felt them. 

I do see an [old slack post from me about me about trying to install clojure core team cli](https://clojurians.slack.com/archives/CFN4QDHPS/p1596059451167400):

> For what it is worth, I had a pleasant experience experimenting with [@borkdude](https://clojurians.slack.com/team/U04V15CAJ)’s [deps.clj](https://github.com/borkdude/deps.clj) (using its deps.exe) when setting up testing of rewrite-cljc on GitHub Actions on Windows. [I was even bold enough to rename](https://github.com/lread/rewrite-cljc-playground/blob/c75cab69b5d48bd87dc1c3e14ce85b07034346ef/.github/workflows/unit-test.yml#L83) deps.exe [ to](https://github.com/lread/rewrite-cljc-playground/blob/9b0cce953df705115c1fc6d4f488a0b3293cecad/.github/workflows/unit-test.yml#L83) `clojure.exe`. I did initially try installing the official [alpha Clojure PowerShell module](https://github.com/clojure/tools.deps.alpha/wiki/clj-on-Windows), but GitHub Actions did not see the module, so I gave up, admittedly, rather quickly. Using a renamed deps.clj allowed me the luxury of calling clojure easily from any shell.

I think you've solved the installation part of this problem with your setup-clojure Action, but not the easily "runnable from any shell" part.

If you want a blurb for the README, maybe something like?:

**cmd-exe-workaround**

Some people bump into challenges using the clojure core team alpha `cli` on Windows because it is runnable only from Powershell. Launching a Powershell module from a script can make such things as argument escaping and exit code capturing extra tricky. For these people, we offer `cmd-exe-workaround` to instead install [borkdude/deps.clj](https://github.com/borkdude/deps.clj)'s `deps.exe` as `clojure.exe` when on Windows.

**Comment by DeLaGuardo** [comment]
thanks!

**Comment by lread** [comment]
Afterthought: `cmd-exe-workaround` might be a bit of an odd name with the direction we seem to be going in?

Maybe: `deps-exe-cli` instead?

**Comment by borkdude** [comment]
or `deps.clj`?

**Comment by lread** [comment]
I was thinking `exe` to hint it would be only active on Windows.
And `cli` to hint that it overrides any `cli` specification.
But if this holds true, maybe there is a clearer way to represent this.

Is it shortsighted of me to assume deps.clj only on Windows? Would someone find a case where they wanted to use a `deps.clj` binary on another OS?

**Comment by lread** [comment]
> Is it shortsighted of me to assume deps.clj only on Windows? Would someone find a case where they wanted to use a deps.clj binary on another OS?

On Slack @borkdude mentioned:
> FWIW I never had anyone ask me to provide a deps.exe for a different architecture

**Comment by borkdude** [comment]
In that context, I meant Windows: not ever has someone request e.g. an aarch64 binary for deps.exe on Windows.

**Comment by lread** [comment]
Ah, thanks! I wrongly assumed you meant that folks only used the deps binary on Windows.

**Comment by borkdude** [comment]
Well, I think that is mostly the case since deps.clj as a binary it solves a problem on Windows mostly.

**Comment by borkdude** [comment]
I think you could now try clj-msi as the workaround on Windows:

https://github.com/casselc/clj-msi/releases

This packages deps.clj as `clj.exe` and `clojure.exe` and installs it on your PATH. This might become the official installer for Clojure on Windows in the future.

**Comment by lread** [comment]
Update: [clj-msi has become the official way to install Clojure on Windows](https://clojure.org/guides/install_clojure#_windows_installer). There should be no need to install the old PowerShell modules anymore.

I'll rename this issue accordingly.

**Comment by lread** [comment]
@DeLaGuardo I could attempt a PR to move to the new official `clojure` on Windows - if you are agreeable to that.

**Comment by DeLaGuardo** [comment]
Fixed in #114

**Comment by borkdude** [comment]
Thank you!
